pyro/contrib/gp/parameterized.py

```diff
@@ -65,70 +63,15 @@ class Parameterized(nn.Module):
     """
     def __init__(self):
         super(Parameterized, self).__init__()
-        self._constraints = OrderedDict()
         self._priors = OrderedDict()
         self._guides = OrderedDict()
         self._mode = None
 
-    def set_constraint(self, name, constraint):
-        """
-        Sets the constraint of an existing parameter.
-
-        :param str name: Name of the parameter.
-        :param ~constraints.Constraint constraint: A PyTorch constraint. See
-            :mod:`torch.distributions.constraints` for a list of constraints.
-        """
-        if constraint in [constraints.real, constraints.real_vector]:
-            if name in self._constraints:  # delete previous constraints
-                self._constraints.pop(name, None)
-                self._parameters.pop("{}_unconstrained".format(name))
-
-                if name not in self._priors:
-                    # no prior -> no guide
-                    # so we can move param back from buffer
-                    p = Parameter(self._buffers.pop(name).detach())
-                    self.register_parameter(name, p)
-            return
-
-        if name in self._priors:
-            raise ValueError("Parameter {} already has a prior. Can not set a constraint for it."
-                             .format(name))
-
-        if name in self._parameters:
-            p = self._parameters.pop(name)
-        elif name in self._buffers:
-            p = self._buffers[name]
-        else:
-            raise ValueError("There is no parameter with name: {}".format(name))
-
-        p_unconstrained = Parameter(transform_to(constraint).inv(p).detach())
-        self.register_parameter("{}_unconstrained".format(name), p_unconstrained)
-        # due to precision issue, we might get f(f^-1(x)) != x
-        # so it is necessary to transform back
-        p = transform_to(constraint)(p_unconstrained)
-        self.register_buffer(name, p.detach())
-        self._constraints[name] = constraint
-
-    def set_prior(self, name, prior):
-        """
-        Sets the constraint of an existing parameter.
-
-        :param str name: Name of the parameter.
-        :param ~pyro.distributions.distribution.Distribution prior: A Pyro prior
-            distribution.
-        """
-        if name in self._parameters:
-            # move param to _buffers
-            p = self._parameters.pop(name)
-            self.register_buffer(name, p)
-        elif name not in self._buffers:
-            raise ValueError("There is no parameter with name: {}".format(name))
-
-        self._priors[name] = prior
-        # remove the constraint and its unconstrained parameter
-        self.set_constraint(name, constraints.real)
-
-        self.autoguide(name, dist.Delta)
+    def __setattr__(self, name, value):
+        super().__setattr__(name, value)
+        if isinstance(value, PyroSample):
```

Comment:
Regarding mode, could pass a state-dependent distribution, maybe like:
```py
def __setattr__(self, name, value):
    if isinstance(value, PyroSample):
        prior = value.prior
        self._priors[name] = prior
        self.autoguide(name, dist.Delta)
        value = PyroSample(lamba self: prior if self.mode == "model" else
                           self.self._get_guide(name.split(".")[-1], name))
    super().__setattr__(name, value)
```

Reply:
nit: why do you call `super().__setattr__()` twice? I would have thought a single `super.__setattr__()` outside the if statement at the end of the method would suffice.

Reply:
My intention was to setup the initial value for autoguide's parameter. But it is not necessary. I'll change it back. Thanks for the trick! ;)